core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,53 +1463,128 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
+         KeyValuePair<Object, FileEntry> next = it.next();
+         FileEntry fe = next.getValue();
+         if (fe.isExpired(now)) {
+            it.set(null);
+
+            // Safe to unlock because the entry was locked collectExpiredEntries
+            // FIXME We cannot load the metadata here or the lifespan check in CallInterceptor will fail
+            MarshallableEntry<K, V> entry = readFromDisk(fe, next.getKey(), false, false);
+            processor.onNext(entry);
+
+            try {
+               free(fe);
+            } catch (Exception e) {
+               throw new PersistenceException(e);
+            }
+         }
+      }
+   }
+
+   @GuardedBy("resizeLock")
+   private List<KeyValuePair<Object, FileEntry>> collectExpiredEntries(long now, Map<K, FileEntry> segmentEntries) {
       List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
-      synchronized (entries) {
-         for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
+      synchronized (segmentEntries) {
+         for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
             Map.Entry<K, FileEntry> next = it.next();
             FileEntry fe = next.getValue();
             if (fe.isExpired(now)) {
                it.remove();
+               // We don't have to worry about other operations freeing the entry while we are reading it,
+               // but we have to lock because readFromDisk() unlocks
+               fe.lock();
                entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
             }
          }
       }
+      return entriesToPurge;
+   }
+
+   @Override
+   public CompletionStage<Long> size(IntSet segments) {
+      return Flowable.fromPublisher(publishKeys(segments, null)).count().toCompletionStage();
+   }
 
+   @Override
+   public CompletionStage<Long> approximateSize(IntSet segments) {
+      if (!segmented && segments.size() < ctx.getCache().getCacheConfiguration().clustering().hash().numSegments()) {
+         return size(segments);
+      }
+      return blockingManager.supplyBlocking(() -> blockingApproximateSize(segments), "sfs-approximateSize");
+   }
+
+   private long blockingApproximateSize(IntSet segments) {
       long stamp = resizeLock.readLock();
+      long size = 0;
       try {
-         for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
-            KeyValuePair<Object, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               try {
-                  free(fe);
-               } catch (Exception e) {
-                  throw new PersistenceException(e);
-               }
-               if (task != null) task.entryPurged(next.getKey());
-            }
+         if (!segmented) {
+            segments = IntSets.immutableSet(0);
```

Comment:
```suggestion
            return getSegmentEntries(0).size();
```

But I am fine with the way it is.